crates/rune/src/runtime/vm_diagnostics.rs

```diff
@@ -0,0 +1,108 @@
+use core::ptr::NonNull;
+
+use ::rust_alloc::sync::Arc;
+
+use rune_alloc::clone::TryClone;
+
+use crate::hash::Hash;
+use crate::runtime::{RuntimeContext, VmResult};
+use crate::{Diagnostics, Unit};
+
+pub trait VmDiagnostics {
+    fn function_used(
+        &mut self,
+        context: &Arc<RuntimeContext>,
+        unit: &Arc<Unit>,
+        hash: Hash,
+        at: usize,
+    ) -> VmResult<()>;
+
+    fn vtable(&mut self) -> &'static VmDiagnosticsObjVtable;
+}
+
+impl VmDiagnostics for Diagnostics {
+    fn function_used(
+        &mut self,
+        context: &Arc<RuntimeContext>,
+        unit: &Arc<Unit>,
+        hash: Hash,
+        at: usize,
+    ) -> VmResult<()> {
```

Comment:
This interface seems more complicated than it has to be.

In my mind all we need to emit is a `hash` and an instruction pointer `at`. Correlating the diagnostic can be done done external to the trait implementation.

Reply:
I removed the `context` and I totally agree with you on this. The `unit` would be handy, to be able to create a warning, which has the span directly attached to it and stay compatible with the other warnings.
If you still prefer to remove the unit here, I would make the warning with a `Option<Span>` and in addition `Option<usize>` for the ip location.
The only problem I see then, is how I can handle this https://github.com/rune-rs/rune/blob/8e84d521e9146e45639127f048cd029a7b7b5a59/crates/rune/src/diagnostics/warning.rs#L49 when no span is defined....